src/csharp/Grpc.Tools.CSharpGenerator/Program.cs

```diff
@@ -0,0 +1,112 @@
+                // This is a hacktastic way of getting sysname from uname ()
+                if (uname(buf) == 0)
+                {
+                    var os = Marshal.PtrToStringAnsi(buf);
+                    if (os == "Darwin")
+                        return true;
+                }
+            }
+            catch
+            {
+            }
+            finally
+            {
+                if (buf != IntPtr.Zero)
+                    Marshal.FreeHGlobal(buf);
+            }
+            return false;
+        }        
+        static void Main(string[] args)
+        {
+            try
+            {
+                var projectDir = args[0];
+                var intermediateOutputDir = args[1];
+
+                var osName = string.Empty;
+
+                if (Path.DirectorySeparatorChar == '\\')
+                    osName = "windows_{0}";
+                else if (IsRunningOnMac())
+                    osName = "macosx_{0}";
+                else if (Environment.OSVersion.Platform == PlatformID.Unix)
+                    osName = "linux_{0}";
+                else
+                {
+                    Console.Error.WriteLine("Could not detect your operating system. The only supported operating systems are Windows, Mac OSX and Linux");
+                    return;
+                }
+
+                var bitness = string.Empty;
+                if (IntPtr.Size == 8)
+                    bitness = "x64";
+                else if (IntPtr.Size == 4)
+                    bitness = "x86";
+                else
+                {
+                    Console.Error.WriteLine("Could not detect your the architecture of your operating system. Only x86 and x64 are currently supported");
+                    return;
+                }
+
+                var exeDir = Path.GetDirectoryName(System.Reflection.Assembly.GetExecutingAssembly().Location);
+                var protocPath = Path.Combine(exeDir.ParentDirectory(), "tools", string.Format(osName, bitness), osName.StartsWith("windows") ? "protoc.exe" : "protoc");
+                var grpcCSharpPluginPath = Path.Combine(exeDir.ParentDirectory(), "tools", string.Format(osName, bitness), osName.StartsWith("windows") ? "grpc_csharp_plugin.exe" : "grpc_csharp_plugin");
+                var protosIncludeDir = projectDir;
+                var outputDir = Path.Combine(projectDir, intermediateOutputDir);
+
+                var inputFiles = from protofile in args.Skip(2).TakeWhile(s => true)
+                                 select Path.Combine(projectDir, protofile);
+                foreach (var f in inputFiles)
+                    Console.WriteLine("Going to process: {0}", f);
+
+                foreach (var inputFile in inputFiles)
+                {
+                    var protocArguments = string.Format("-I{0} --csharp_out {1} --grpc_out {1} --plugin=protoc-gen-grpc={2} {3}", protosIncludeDir, outputDir, grpcCSharpPluginPath, inputFile);

```

Comment:
As a first step, I can support this using the "Custom Tool namespace" metadata on that item. 

Later, I can extend it to be generic by option name. See this example of custom metadata [here](https://msdn.microsoft.com/en-us/library/ms171474.aspx). Consumers that want to use this advanced feature will have to hand-edit their project files and add the metadata in.

Does this sound good?